source/uwp/SharedRenderer/lib/Util.cpp

```diff
@@ -535,19 +536,50 @@ bool IsBackgroundImageValid(winrt::AdaptiveBackgroundImage const& backgroundImag
     return false;
 }
 
+winrt::Uri UriTryCreate(winrt::hstring const &uri, winrt::hstring const& baseUri = L"")
+{
+    auto factory = winrt::
+        get_activation_factory<winrt::Windows::Foundation::Uri, winrt::Windows::Foundation::IUriRuntimeClassFactory>();
+    auto abiFactory = static_cast<ABI::Windows::Foundation::IUriRuntimeClassFactory *>(winrt::get_abi(factory));
+
+    const winrt::hstring &localUri = uri;
+    winrt::Windows::Foundation::Uri returnValue{nullptr};
+    HRESULT hr = S_OK;
+    if (baseUri.empty())
+    {
+        hr = abiFactory->CreateUri(
+            static_cast<HSTRING>(winrt::get_abi(localUri)),
+            reinterpret_cast<ABI::Windows::Foundation::IUriRuntimeClass**>(winrt::put_abi(returnValue)));
+    }
+    else
+    {
+        hr = abiFactory->CreateWithRelativeUri(
+            static_cast<HSTRING>(winrt::get_abi(baseUri)),
+            static_cast<HSTRING>(winrt::get_abi(uri)),
+            reinterpret_cast<ABI::Windows::Foundation::IUriRuntimeClass**>(winrt::put_abi(returnValue)));
+    }
+
+    if (FAILED(hr))
+    {
+        return winrt::Windows::Foundation::Uri { nullptr };
+    }
+
+    return returnValue;
+}
+
 winrt::Uri GetUrlFromString(winrt::AdaptiveHostConfig const& hostConfig, winrt::hstring const& urlString)
 {
     winrt::Uri uri{nullptr};
 
-    if (const auto uriFromAbsoluteUri = winrt::Uri{urlString})
+    if (const auto uriFromAbsoluteUri = UriTryCreate(urlString))
     {
         return uriFromAbsoluteUri;
     }
     else
     {
         winrt::hstring imageBaseUrl = hostConfig.ImageBaseUrl();
 
-        if (const auto uriFromRelativeUri = winrt::Uri{imageBaseUrl, urlString})
+        if (const auto uriFromRelativeUri = UriTryCreate(urlString, imageBaseUrl))
         {
             return uriFromRelativeUri;
         }
```

Comment:
do you think it could be better to throw, then returning the nullptr? I think that we catch exceptions from rendering. if it's too much work, i don't think it's worth it, but if the catch block is there, then the error handling code would be simpler and slightly better.

Reply:
I meant at LN 587

Reply:
I prefer exceptions for cases that more fatal in nature and they lead to program termination or some fundamental routine exit. In this case, we're simply parsing a `URI` and throwing an exception (which will capture entire stack as well) looks like an overkill to me :)
In some of the places where we've been using `GetUrlFromString` like here: https://github.com/microsoft/AdaptiveCards/blob/4745cb65ae8d57529811f7a0ba13f04afbadcacc/source/uwp/SharedRenderer/lib/AdaptiveImageRenderer.cpp#L57-L59 -> we already expect it to return `nullptr` on failure. Before this PR, this would never happen and the method would throw, leading to exception bubbling up and we wouldn't know what's wrong. In this case, it will work as expected.

Reply:
that's fine. all i'm saying is if we throw, the exception will get caught by elemental renderer, and will get skipped which is the desired behavior.